### PR TITLE
Mutualize MYSQL connection in case of MySQL dictionaries

### PR DESCRIPTION
**Pull request by clemrodriguez**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Improvement

Short description:
For external MySQL dictionaries, mutualize MySQL connection pool to "share" them among dictionaries.
The aim is to reduce total number of MySQL connections.
It is related to issue #893 and goes forward pull request #5395

Detailed description:
In our context we use many External MySQL dictionaries, they are not frequently updated. As each connection consumes resources on the MySQL server side, we would like to reduce the total number of opened ones. So we would like to mutualize them, the pool will re-use MySQL closed connections to feed connection consumers (External Dictionaries).

To do so:
- we added a bool parameter share_connection in external dictionary definition dictionary.xml (at same level as db or table), if its value is true it forces connection closure after usage
- modified MySQLDictionarySource to consider the dictionary parameter (share_connection) with default value = false (previous behavior) and force close connection if set to true. It prevents connection famine pattern.
- create a class called PoolFactory in libmysqlxx. It is a singleton storing PoolWithFailOver associated to a dictionary configuration. If a dictionary is declared having a shareable connection, then this class will generate a "connection name" based on "user@host:port/db".
- if an external dictionary request similar connection they will share the pool
- modified ExternalDictionaries class to reset PoolFactory map when receiving SYSTEM RELOAD DICTIONARY | DICTIONARIES

### CODE REVIEW

File: libs/libmysqlxx/include/mysqlxx/PoolFactory.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#define MYSQLXX_POOL_WITH_FAILOVER_DEFAULT_MAX_CONNECTIONS 16
+#define MYSQLXX_POOL_WITH_FAILOVER_DEFAULT_MAX_TRIES 3
+
+namespace mysqlxx
+{
+/*
+ * PoolFactory.h
+ * This class is a helper singleton to mutualize connections to MySQL.
+ */
+class PoolFactory final : public ext::singleton<PoolFactory>
```
Comment (alexey-milovidov) [inline libs/libmysqlxx/include/mysqlxx/PoolFactory.h]: PoolFactory must be made thread-safe even if currently its usage is always serialized.

**Comment by alexey-milovidov** [comment]
We need an integration test where the new code at least runs.

**Comment by alexey-milovidov** [comment]
Do you want to add an integration test?

**Comment by clemrodriguez** [comment]
Sorry for the delay. To be honest I wouldn't mind to add integration test. But I don't know how to make them.
Modified code is executed by 'system reload dictionaries' or any call on a dictionary.

**Comment by alexey-milovidov** [comment]
> But I don't know how to make them.

Integration tests are located in `dbms/tests/integration`.
You can look for similar tests.

**Comment by clemrodriguez** [comment]
@alexey-milovidov 
Hi Alexey,
I made a forced push on my local branch from my fork (which wasn't the pull request branch). Just after that I saw that the PR is closed.
Capture
As I'm not used to Git, I'm upset of having pushed things that wasn't expected.
Can you confirm me it had not impact on your repo and it just closed the PR ?
Due to a recent rebase, my commit log is really poluted and I wanted to reopen a new one for the same topic.

Best regards.

**Comment by alexey-milovidov** [comment]
Most likely, you have accidentially force-pushed upstream master code to your PR and now it has no difference to upstream (0 files changed). GitHub has automatically closed this PR because it is empty. (Don't worry, nothing wrong happens to our repo).

Do you have your code in your branch? It's better to create another PR.

BTW, this is important feature and we need it also in our company.

**Comment by clemrodriguez** [comment]
Thank you for the your feedback.
Code is integrated into another local branch (v19.15.7.2) as we need in our production environment.

I'll close this PR and open a new one.

**Pull request closed without merging**